### PR TITLE
Capacity staking min token balance

### LINKED ISSUE #944: [Capacity] staking should not consume the entire free balance
Currently, a user is allowed to stake all of their free balance as this could be inconvenient when you unstake and are left with 0 balance. 
We would like this Minimum Token Balance to be a configurable amount with a minimum value that must be greater than the Existential Deposit.

AC:
- [ ] The MinimumTokenBalance should be deducted from the staked amount when the user attempts to stake their entire free balance.
- [ ] Stake transactions should always leave the remaining token balance >= MinimumTokenBalance.
- [ ] An error should be thrown when the total stake amount would result in a token balance < MinimumTokenBalance.

Case Scenarios
- [x] Scenario A:
1. Bob has a balance of 10 tokens.
2. Bob stakes all of his tokens.
3. Result: only 9 tokens have been staked because the MinimumTokenBalance is 1.
- [x] Scenario B:
1. Bob has a balance of 1 token.
2. Bob stakes all of his tokens.
3. Result: Error should occur because staking 1 token does not leave any tokens available to meet the requirements of the MinimumTokenBalance = 1.
- [x] Scenario C:
1. Bob has a balance of 10 tokens.
2. Bob stakes 3 tokens.
3. Result: 3 tokens successfully staked: the remaining token balance is greater than the MinimumTokenBalance.

### PR DESCRIPTION
**Pull request by mattheworris**
# Goal
The goal of this PR is to implement a Minimum Token Balance. 
Currently, a user is allowed to stake all of their free balance. 
This could be inconvenient when you unstake and are left with 0 balance. 
This Minimum Token Balance is a configurable amount with a minimum value that must be greater than the Existential Deposit.

Closes #944 

# Checklist
- [x] Design doc(s) updated
- [x] Tests added
- [x] Weights updated
# N/A
- [ ] Chain spec updated
- [ ] Custom RPC OR Runtime API added/changed? Updated js/api-augment.
- [ ] Benchmarks added

### CODE REVIEW

File: pallets/capacity/src/weights.rs [hunk 1]
```diff
@@ -18,11 +18,11 @@
 //! Autogenerated weights for pallet_capacity
 //!
 //! THIS FILE WAS AUTO-GENERATED USING THE SUBSTRATE BENCHMARK CLI VERSION 4.0.0-dev
-//! DATE: 2023-02-03, STEPS: `20`, REPEAT: 10, LOW RANGE: `[]`, HIGH RANGE: `[]`
+//! DATE: 2023-03-17, STEPS: `50`, REPEAT: 10, LOW RANGE: `[]`, HIGH RANGE: `[]`
```
Comment (enddynayn) [inline pallets/capacity/src/weights.rs:21]: 💯

**Review by enddynayn — APPROVED** [review]
👍 Nice! Belleza!

File: pallets/capacity/src/lib.rs [hunk 2]
```diff
@@ -253,7 +258,7 @@ pub mod pallet {
 	pub enum Error<T> {
 		/// Staker attempted to stake to an invalid staking target.
 		InvalidTarget,
-		/// Staker has insufficient balance to cover the amount wanting to stake.
+		/// Capacity is not available for the given MSA.
```
Comment (johnmela) [inline pallets/capacity/src/lib.rs:261]: 💯

**Review by johnmela — APPROVED** [review]

File: integration-tests/capacity/replenishment.test.ts [hunk 3]
```diff
@@ -7,15 +7,17 @@ import {
...
+  TEST_EPOCH_LENGTH
 } from "../scaffolding/helpers";
 import { firstValueFrom} from "rxjs";
 import {AVRO_GRAPH_CHANGE} from "../schemas/fixtures/avroGraphChangeSchemaType";
 
 
 describe("Capacity Replenishment Testing: ", function () {
-  const TEST_EPOCH_LENGTH=10;
-  const STARTING_BALANCE=6n* 1000n * 1000n;
+  const STARTING_BALANCE=6n* 1000n * 1000n + 100n*1000n*1000n;
```
Comment (shannonwells) [inline integration-tests/capacity/replenishment.test.ts:20]: ?
```suggestion
  const STARTING_BALANCE= 106n* 1000n * 1000n;
```

File: integration-tests/capacity/replenishment.test.ts [hunk 4]
```diff
@@ -126,7 +123,7 @@ describe("Capacity Replenishment Testing: ", function () {
       const [stakeKeys, stakeProviderId] = await createAndStakeProvider("TinyStake", 1n*1000n*1000n);
       // new user/msa stakes to provider
       const userKeys = createKeys("userKeys");
-      await fundKeypair(devAccounts[0].keys, userKeys, 2n*1000n*1000n);
+      await fundKeypair(devAccounts[0].keys, userKeys, 2n*1000n*1000n + 100n*1000n*1000n);
```
Comment (shannonwells) [inline integration-tests/capacity/replenishment.test.ts:126]: a little easier to read
```suggestion
      await fundKeypair(devAccounts[0].keys, userKeys, 102n*1000n*1000n);
```

File: integration-tests/capacity/transactions.test.ts [hunk 5]
```diff
@@ -25,7 +25,7 @@ describe("Capacity Transaction Tests", function () {
     let delegatorProviderId: u64;
     let schemaId: u16;
 
-    let stakeAmount: bigint = 20000000n;
+    let stakeAmount: bigint = 20n* 1000n* 1000n* 1000n;
```
Comment (shannonwells) [inline integration-tests/capacity/transactions.test.ts:28]: this is 1000X more than 20,000,000n - is that on purpose?

File: pallets/capacity/src/tests/staking_account_details_tests.rs [hunk 6]
```diff
@@ -116,7 +116,7 @@ fn impl_staking_account_details_default() {
 #[test]
 fn impl_staking_account_details_get_stakable_amount_for() {
 	new_test_ext().execute_with(|| {
-		let account = 100;
+		let account = 200;
```
Comment (shannonwells) [inline pallets/capacity/src/tests/staking_account_details_tests.rs:119]: Why this change?

**Review by shannonwells — APPROVED** [review]
Nothing blocking, but I think there are some unnecessary changes.

**Comment by enddynayn** [comment]
I'll be addressing these comments on a refactoring PR.
